python/pyspark/sql/types.py

```diff
@@ -189,7 +189,7 @@ def toInternal(self, dt):
         if dt is not None:
             seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
```

Comment:
https://github.com/bartdag/py4j/blob/master/py4j-python/src/py4j/protocol.py#L271-L275

Py4J automatically serializes any Python integer larger than 2 ^ 31 as `LONG_TYPE`, otherwise it's `INTEGER_TYPE`. Python longs are always serialized as `LONG_TYPE`.

I suspect my issue with Python 3 is that there is no more `long`, it's all just `int`. This may require a fix on the Scala side to accept either an int or a long to the appropriate constructor.